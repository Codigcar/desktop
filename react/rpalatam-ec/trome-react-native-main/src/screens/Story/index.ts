import StoryScreen from './Story'

export default StoryScreen
