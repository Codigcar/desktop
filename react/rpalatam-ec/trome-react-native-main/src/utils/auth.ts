import { appleAuth } from '@invertase/react-native-apple-authentication'
import { GoogleSignin } from '@react-native-community/google-signin'
import axios from 'axios'
import { Linking } from 'react-native'
import NativeConfig from 'react-native-config'
import { AccessToken, LoginManager } from 'react-native-fbsdk'
import { InAppBrowser } from 'react-native-inappbrowser-reborn'
import { randomBytes } from 'react-native-randombytes'
import { sha256 } from 'react-native-sha256'
import { URL, URLSearchParams } from 'react-native-url-polyfill'

import type { OAuth, Token } from '../services/auth/auth.types'

export async function appleSignIn(): Promise<string> {
  const appleAuthRequestResponse = await appleAuth.performRequest({
    requestedOperation: appleAuth.Operation.LOGIN,
    // Note: it appears putting FULL_NAME first is important, see issue #293
    requestedScopes: [appleAuth.Scope.FULL_NAME, appleAuth.Scope.EMAIL],
  })

  if (!appleAuthRequestResponse.identityToken) {
    throw new Error('Without id token')
  }

  const credentialState = await appleAuth.getCredentialStateForUser(
    appleAuthRequestResponse.user,
  )

  if (credentialState !== appleAuth.State.AUTHORIZED) {
    throw new Error('Without permissions')
  }
  return appleAuthRequestResponse.identityToken
}

export async function googleSignIn(): Promise<string> {
  await GoogleSignin.hasPlayServices()
  const { idToken } = await GoogleSignin.signIn()
  if (!idToken) throw new Error('Without id token')
  return idToken
}

export async function googleSignOut(): Promise<void> {
  try {
    await GoogleSignin.revokeAccess()
    await GoogleSignin.signOut()
  } catch (_) {}
}

export async function facebookSignIn(): Promise<string> {
  LoginManager.logOut()
  const result = await LoginManager.logInWithPermissions([
    'public_profile',
    'email',
  ])
  if (result.isCancelled) throw new Error('Cancelled')
  const data = await AccessToken.getCurrentAccessToken()
  const token = data?.accessToken.toString()
  if (!token) throw new Error('Without token')
  return token
}

export function init(): void {
  const options = {
    webClientId: NativeConfig.GOOGLE_WEB_CLIENT_ID,
    offlineAccess: true,
    iosClientId: NativeConfig.GOOGLE_IOS_CLIENT_ID,
  }
  GoogleSignin.configure(options)
}

function base64URLEncode(input: Buffer) {
  return input
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/[=]/g, '')
}

export const PKCE: Record<string, string | null> = {
  code_verifier: null,
  state: null,
}

interface TokenExchangeResponse {
  access_token: string
  refresh_token: string
  token_type: string
  expires_in: number
}

export async function tokenExchange(
  config: OAuth,
  url: string,
): Promise<Token> {
  const { searchParams } = new URL(url)
  if (searchParams.get('state') !== PKCE.state) {
    throw new Error(`OAuth state isn't found`)
  }
  const params = new URLSearchParams({
    client_id: config.client_id,
    code: String(searchParams.get('code')),
    code_verifier: String(PKCE.code_verifier),
    grant_type: 'authorization_code',
    redirect_uri: config.redirect_url,
  })
  const response = await axios.post<TokenExchangeResponse>(
    `${config.issuer}/token`,
    params.toString(),
    {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    },
  )
  PKCE.code_verifier = null
  PKCE.state = null
  return {
    access_token: response.data.access_token,
    refresh_token: response.data.refresh_token,
  }
}

export async function authorize(config: OAuth): Promise<string | null> {
  PKCE.code_verifier = base64URLEncode(randomBytes(32))
  PKCE.state = base64URLEncode(randomBytes(16))
  const code_encrypted = await sha256(PKCE.code_verifier)
  const challenge = base64URLEncode(Buffer.from(code_encrypted, 'hex'))
  const authorizeParams = new URLSearchParams({
    ...config.additional_parameters,
    client_id: config.client_id,
    code_challenge: challenge,
    code_challenge_method: 'S256',
    redirect_uri: config.redirect_url,
    response_type: 'code',
    state: PKCE.state,
  })
  const url = `${config.issuer}/authorize?${authorizeParams.toString()}`
  if (!(await InAppBrowser.isAvailable())) {
    await Linking.openURL(url)
    return null
  }
  const result = await InAppBrowser.openAuth(
    url,
    encodeURIComponent(config.redirect_url),
  )
  if (result.type !== 'success') return null
  return result.url
}
