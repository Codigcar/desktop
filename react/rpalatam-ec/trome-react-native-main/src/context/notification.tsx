import React, {
  useCallback,
  useContext,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react'
import { Animated, StyleSheet } from 'react-native'
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'

import Box from '../components/box'
import Typography from '../components/typography'
import { App } from '../utils/config'

const { SafeAreaView, View } = Box
const { Paragraph } = Typography

interface Context {
  hide: () => void
  show: (opts: NotifyOptions) => void
}

type NotifyType = 'success' | 'error'

export interface NotifyOptions {
  message: string
  type: NotifyType
  duration?: number
  action?: React.ComponentType | null
}

interface NotifyProps {
  close(callback: () => void): void
  open(): void
}

type NotifyRef = {
  update(options?: NotifyOptions): void
  open: boolean
}

export const NotificationContext = React.createContext<Context>({} as Context)

const bgcolor: Record<NotifyType, string> = {
  success: App.select({
    gestion: '#1781a8',
    default: '#0089ff',
  }),
  error: '#EF4444',
}

const icon: Record<NotifyType, string> = {
  success: 'check-circle-outline',
  error: 'close-circle-outline',
}

const Notify = React.forwardRef<NotifyRef, NotifyProps>((props, ref) => {
  const closeTimeoutRef = useRef<ReturnType<typeof setTimeout> | null>(null)
  const [notify, setNotify] = useState<NotifyOptions>()
  const { close, open } = props

  useEffect(() => {
    if (!!notify) {
      open()
      closeTimeoutRef.current = setTimeout(() => {
        close(() => setNotify(undefined))
      }, notify.duration || 5000)
    }
    return () => {
      closeTimeoutRef.current && clearTimeout(closeTimeoutRef.current)
    }
  }, [close, open, notify])

  useImperativeHandle(
    ref,
    () => ({
      update: setNotify,
      open: false,
    }),
    [],
  )

  if (!notify) return null

  const { message, type, action: Action } = notify

  return (
    <View style={{ backgroundColor: bgcolor[type] }}>
      <SafeAreaView edges={['bottom']}>
        <View alignItems="center" flexDirection="row" px="0.5" py="0.75">
          <View flex={1} alignItems="center" flexDirection="row">
            <View mr="0.5">
              <Icon name={icon[type]} color="#FFFFFF" size={16} />
            </View>
            <View flex={1}>
              <Paragraph color="white" fontSize="sm">
                {message}
              </Paragraph>
            </View>
          </View>
          {!Action ? null : <Action />}
        </View>
      </SafeAreaView>
    </View>
  )
})

export const NotificationProvider: React.FC = ({ children }) => {
  const notifyRef = useRef<NotifyRef>(null)
  const translateAnimatedRef = useRef(new Animated.Value(0))
  const translateValue = translateAnimatedRef.current.interpolate({
    inputRange: [0, 1],
    outputRange: [200, 0],
  })

  const animatedStyles = {
    transform: [
      {
        translateY: translateValue,
      },
    ],
  }

  const handleClose = useCallback((callback?: () => void) => {
    Animated.timing(translateAnimatedRef.current, {
      toValue: 0,
      useNativeDriver: true,
      duration: notifyRef.current?.open ? 250 : 0,
    }).start(() => {
      if (notifyRef.current) notifyRef.current.open = false
      callback?.()
    })
  }, [])

  const handleOpen = useCallback(() => {
    Animated.timing(translateAnimatedRef.current, {
      toValue: 1,
      useNativeDriver: true,
      duration: 250,
    }).start()
    if (notifyRef.current) notifyRef.current.open = true
  }, [])

  const hide = useCallback(() => {
    handleClose(() => notifyRef.current?.update())
  }, [handleClose])

  const show = useCallback(
    (options: NotifyOptions) => {
      handleClose(() => notifyRef.current?.update(options))
    },
    [handleClose],
  )

  return (
    <NotificationContext.Provider value={{ hide, show }}>
      {children}
      <Animated.View style={[styles.container, animatedStyles]}>
        <Notify ref={notifyRef} close={handleClose} open={handleOpen} />
      </Animated.View>
    </NotificationContext.Provider>
  )
}

export const useNotification = (): Context => useContext(NotificationContext)

const styles = StyleSheet.create({
  container: {
    bottom: 0,
    left: 0,
    position: 'absolute',
    transform: [{ translateY: 200 }],
    width: '100%',
  },
})
