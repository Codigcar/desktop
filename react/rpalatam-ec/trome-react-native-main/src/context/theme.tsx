import AsyncStorage from '@react-native-async-storage/async-storage'
import { ThemeProvider as SystemProvider } from '@shopify/restyle'
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react'
import { Appearance, Platform } from 'react-native'

import lightTheme, { darkTheme } from '../theme'
import { STORE_SCHEME } from '../utils/constants'
import { HomeWebviewRef } from '../utils/refs'

export type ThemeKey = 'light' | 'dark'
type SchemeKey = ThemeKey | 'system'
interface ThemeContextInterface {
  currentTheme: ThemeKey
  schemeSelection: SchemeKey
  sendThemeToWebview: () => void
  updateSchemeSelection: (key: SchemeKey) => void
}

export const ThemeContext = createContext({} as ThemeContextInterface)

export const darkModeIsSupported =
  (Platform.OS === 'android' && Platform.Version >= 29) ||
  (Platform.OS === 'ios' && parseInt(Platform.Version as string, 10) >= 13)

export const INITIAL_THEME: ThemeKey = 'light'

export const ThemeProvider: React.FC = ({ children }) => {
  const [colorScheme, setColorScheme] = useState(
    Appearance.getColorScheme() || INITIAL_THEME,
  )
  const [schemeSelection, setSchemeSelection] = useState<SchemeKey>(
    darkModeIsSupported ? 'system' : colorScheme,
  )

  const currentTheme = useMemo(() => {
    if (schemeSelection === 'system') return colorScheme
    return schemeSelection
  }, [colorScheme, schemeSelection])

  const appTheme = useMemo(
    () => (currentTheme === 'dark' ? darkTheme : lightTheme),
    [currentTheme],
  )

  const sendThemeToWebview = useCallback(() => {
    HomeWebviewRef.current?.injectJavaScript(`(function(){
      if (window.NATIVE_CONNECTION && window.NATIVE_CONNECTION.mode) {
        window.NATIVE_CONNECTION.mode.setNightMode(${currentTheme === 'dark'})
      }
    })()`)
  }, [currentTheme])

  const updateSchemeSelection = useCallback((value: SchemeKey) => {
    AsyncStorage.setItem(STORE_SCHEME, value)
    setSchemeSelection(value)
  }, [])

  const timeout = useRef<ReturnType<typeof setTimeout> | null>(null)
  useEffect(() => {
    const onColorSchemeChange = (
      preferences: Appearance.AppearancePreferences,
    ) => {
      const color = preferences.colorScheme
      if (timeout.current) clearTimeout(timeout.current)
      timeout.current = setTimeout(() => {
        if (color) setColorScheme(color)
      }, 250)
    }
    const subscription = Appearance.addChangeListener(onColorSchemeChange)
    return () => {
      if (timeout.current) clearTimeout(timeout.current)
      subscription.remove()
    }
  }, [])

  useEffect(() => {
    AsyncStorage.getItem(STORE_SCHEME).then((theme) => {
      if (!theme) return
      setSchemeSelection(theme as SchemeKey)
    })
  }, [])

  useEffect(() => {
    sendThemeToWebview()
  }, [sendThemeToWebview])

  return (
    <ThemeContext.Provider
      value={{
        currentTheme,
        schemeSelection,
        sendThemeToWebview,
        updateSchemeSelection,
      }}>
      <SystemProvider theme={appTheme}>{children}</SystemProvider>
    </ThemeContext.Provider>
  )
}

export const useThemeContext = (): ThemeContextInterface =>
  useContext(ThemeContext)
