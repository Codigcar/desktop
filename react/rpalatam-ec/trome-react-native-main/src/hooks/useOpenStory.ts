import { useNavigation, useRoute } from '@react-navigation/native'
import { useCallback, useEffect, useMemo } from 'react'

import { useMainNavigation } from '../context/navigation'
import { Story } from '../entities/Story'
import { ENABLE_SCREEN_STORY } from '../utils/flags'
import history from '../utils/history'
import type { MainStackScreenProps } from '../routes/types'

const ref: { id: string | null; pathname: string | null } = {
  id: null,
  pathname: null,
}

const useOpenStory = (currentScreen: string): ((story: Story) => void) => {
  const navigation = useNavigation<MainStackScreenProps<'Home'>['navigation']>()
  const route = useRoute()
  const { categories } = useMainNavigation()
  const categoriesPath = useMemo(() => {
    return categories.map(({ path }) => path.replace(/^\/category/, ''))
  }, [categories])

  useEffect(() => {
    return navigation.addListener('beforeRemove', () => {
      ref.id = null
      ref.pathname = null
    })
  }, [navigation])

  const main = useCallback(
    (story: Story) => {
      if (ENABLE_SCREEN_STORY) {
        if (ref.id === story.id || ref.pathname === story.url) return
        ref.id = story.id || null
        ref.pathname = story.url || null
        navigation.push('Story', { id: story.id, pathname: story.url })
        return
      }

      const category = categoriesPath.find((path) =>
        story.url?.startsWith(path),
      )
      const prefix = !!category ? '' : '/news'
      const path = `${prefix}${story.url}?ref=${route.name}Screen`
      history.push(path, {
        section: currentScreen,
        pageInfo: {
          screenName: route.name,
          params: route.params,
        },
      })
      navigation.navigate('Main', {
        screen: 'Home',
        params: { screen: 'Feed' },
      })
    },
    [categoriesPath, currentScreen, navigation, route],
  )

  return main
}

export default useOpenStory
