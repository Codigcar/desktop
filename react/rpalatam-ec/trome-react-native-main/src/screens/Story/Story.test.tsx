import {
  useFocusEffect,
  useNavigation,
  useRoute,
} from '@react-navigation/native'
import { act, cleanup, fireEvent, render } from '@testing-utils/library'
import MockAxios from 'jest-mock-axios'
import React from 'react'
import { Alert, Platform, Share } from 'react-native'

import Story from './Story'
import { useAuth } from '../../context/auth'
import { useFavorites } from '../../context/favorites'
import { NotificationProvider } from '../../context/notification'
import useOpenStory from '../../hooks/useOpenStory'

jest.mock('@react-navigation/native')
const mockUseFocusEffect = useFocusEffect as jest.Mock
const mockUseNavigation = useNavigation as jest.Mock
const mockUseRoute = useRoute as jest.Mock

jest.mock('../../context/auth')
const mockUseAuth = useAuth as jest.Mock

jest.mock('../../context/favorites')
const mockUseFavorites = useFavorites as jest.Mock

jest.mock('../../hooks/useOpenStory')
const mockUseOpenStory = useOpenStory as jest.Mock

/**
 * FIXME: Resolve the webview mock ref and remove this spy
 */
jest.spyOn(console, 'error').mockImplementation(() => jest.fn())
jest.useFakeTimers()

beforeEach(() => {
  mockUseAuth.mockImplementation(() => jest.fn())
  mockUseFavorites.mockImplementation(() => ({ favorites: [] }))
  mockUseFocusEffect.mockImplementation((fn) => fn())
  mockUseNavigation.mockImplementation(() => ({ push: jest.fn() }))
  mockUseNavigation.mockImplementation(() => ({
    push: jest.fn(),
    addListener: jest.fn(),
  }))
  mockUseRoute.mockImplementation(() => ({
    params: { id: 'id', pathname: '/notice/' },
  }))
})

afterEach(() => {
  MockAxios.reset()
  cleanup()
})

describe('Story screen', () => {
  it('snapshot', () => {
    const { toJSON } = render(<Story />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('share story', () => {
    Platform.OS = 'ios'
    const spyShare = jest.spyOn(Share, 'share')
    const { getByTestId } = render(<Story />)

    fireEvent.press(getByTestId('button-share'))
    expect(spyShare).toBeCalledWith(
      { url: expect.stringContaining('/notice/') },
      { dialogTitle: 'Compartir' },
    )
  })

  it('open font size modal', async () => {
    const { getByTestId } = render(<Story />)
    fireEvent.press(getByTestId('button-font_size'))
    act(jest.runAllTimers)
    expect(getByTestId('backdrop-touchable')).toBeDefined()
  })

  describe('story params without id or pathname', () => {
    it('story pathname is undefined', () => {
      mockUseRoute.mockImplementationOnce(() => ({ params: { id: 'id' } }))
      const { getByTestId, queryByTestId } = render(<Story />)
      expect(queryByTestId('options')).toBeNull()
      expect(getByTestId('story-webview')).toBeDefined()
    })

    it('story id is undefined', () => {
      mockUseRoute.mockImplementationOnce(() => ({ params: { pathname: '/' } }))
      const { queryByTestId } = render(<Story />)
      expect(queryByTestId('options')).toBeNull()
      expect(queryByTestId('story-webview')).toBeNull()
    })

    it('set navigation screen params', async () => {
      const setParams = jest.fn()
      mockUseNavigation.mockImplementationOnce(() => ({ setParams }))
      mockUseRoute.mockImplementationOnce(() => ({ params: { id: 'id' } }))
      render(<Story />)
      const response = { _id: 'id', canonical_url: '/url/' }
      await act(async () => MockAxios.mockResponse({ data: response }))
      expect(setParams).toBeCalledWith({ id: 'id', pathname: '/url/' })
    })
  })

  describe('handle postmessage', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    function onMessage<T>(data: T): Record<string, any> {
      const utils = render(<Story />)
      const webview = utils.getByTestId('story-webview')
      const elm = webview.parent?.instance._reactInternals._debugOwner
      const injectJavaScript = jest.spyOn(elm.ref.current, 'injectJavaScript')
      webview.props.onMessage({ nativeEvent: { data: JSON.stringify(data) } })
      return { injectJavaScript, props: elm.pendingProps }
    }

    it('should not inject javascript to webview', () => {
      const ref1 = onMessage({ type: 'unknown' })
      expect(ref1.injectJavaScript).not.toBeCalled()
      const ref2 = onMessage(undefined)
      expect(ref2.injectJavaScript).not.toBeCalled()
    })

    it('send user to loaded webview', async () => {
      mockUseAuth.mockImplementation(() => ({
        isAuthenticated: true,
        user: { id: '123' },
      }))
      const { injectJavaScript } = onMessage({ type: 'init.WEB_LOADED' })
      expect(injectJavaScript).toHaveBeenCalledWith(
        expect.stringContaining(JSON.stringify({ id: '123' })),
      )
    })

    it('send theme to loaded webview', () => {
      const { injectJavaScript } = onMessage({ type: 'init.WEB_LOADED' })
      expect(injectJavaScript).toHaveBeenCalledWith(
        expect.stringContaining('mode.setNightMode(false)'),
      )
    })

    it('navigate to previous screen', () => {
      const goBack = jest.fn()
      mockUseNavigation.mockImplementationOnce(() => ({ goBack }))
      onMessage({ type: 'navigation.GO_BACK' })
      expect(goBack).toBeCalled()
    })

    it('navigate to previous screen when not authenticated or signwallModal is active', async () => {
      const goBack = jest.fn()
      mockUseNavigation.mockImplementation(() => ({ goBack }))
      mockUseAuth.mockImplementation(() => ({ isAuthenticated: false }))
      onMessage({
        type: 'navigation.NAVIGATE_TO_SCREEN',
        payload: { name: 'SignwallModal' },
      })

      expect(goBack).toBeCalled()
    })

    it('navigate to screen', () => {
      const navigate = jest.fn()
      mockUseNavigation.mockImplementationOnce(() => ({ navigate }))
      const payload = { name: 'Stack', params: { screen: 'Screen' } }
      onMessage({ type: 'navigation.NAVIGATE_TO_SCREEN', payload })
      expect(navigate).toBeCalledWith(payload.name, payload.params)
    })

    it('navigate to Author screen', () => {
      const navigate = jest.fn()
      mockUseNavigation.mockReturnValueOnce({ navigate })
      const { getByTestId } = render(<Story />)
      const { parent } = getByTestId('story-webview')
      const { onShouldStartLoadWithRequest: fn } =
        parent?.instance._reactInternals._debugOwner.pendingProps
      jest.spyOn(Platform, 'select').mockReturnValueOnce(true)
      const url = 'https://gestion.pe/autor/redaccion-gestion/'
      expect(fn({ url })).toBeFalsy()
      expect(navigate).toBeCalledWith('Author', {
        path: 'redaccion-gestion',
      })
    })

    it('navigate to Tag screen', () => {
      const navigate = jest.fn()
      mockUseNavigation.mockReturnValueOnce({ navigate })
      const { getByTestId } = render(<Story />)
      const { parent } = getByTestId('story-webview')
      const { onShouldStartLoadWithRequest: fn } =
        parent?.instance._reactInternals._debugOwner.pendingProps
      jest.spyOn(Platform, 'select').mockReturnValueOnce(true)
      const url = 'https://gestion.pe/noticias/fitch-ratings/'
      expect(fn({ url })).toBeFalsy()
      expect(navigate).toBeCalledWith('Tag', {
        path: 'fitch-ratings',
      })
    })

    it('navigate to Story screen with useOpenStory', () => {
      const openStory = jest.fn()
      mockUseOpenStory.mockReturnValue(openStory)
      const { getByTestId } = render(<Story />)
      const { parent } = getByTestId('story-webview')
      const { onShouldStartLoadWithRequest: fn } =
        parent?.instance._reactInternals._debugOwner.pendingProps
      jest.spyOn(Platform, 'select').mockReturnValueOnce(true)
      const url = 'https://gestion.pe/economia/peru-noticia/'
      expect(fn({ url })).toBeFalsy()
      expect(openStory).toBeCalledWith({
        id: '',
        url: '/economia/peru-noticia/',
      })
    })

    describe('subscription', () => {
      it('check access for user who is anonymous', () => {
        const navigate = jest.fn()
        mockUseNavigation.mockImplementationOnce(() => ({ navigate }))
        const fnVerifySubscription = jest.fn().mockResolvedValue(null)
        mockUseAuth.mockImplementationOnce(() => ({
          isSubscribed: false,
          user: {},
          verifySubscription: fnVerifySubscription,
        }))

        onMessage({ type: 'subscription.CHECK_ACCESS' })
        expect(navigate).toHaveBeenLastCalledWith('Auth', {
          screen: 'InitialScreen',
        })
      })

      it('check access for user who is subscriber', () => {
        const fnVerifySubscription = jest.fn().mockResolvedValue(null)
        mockUseAuth.mockImplementationOnce(() => ({
          isSubscribed: true,
          user: { id: '123' },
          verifySubscription: fnVerifySubscription,
        }))

        onMessage({ type: 'subscription.CHECK_ACCESS' })
        expect(fnVerifySubscription).toBeCalled()
      })
    })
  })

  describe('read later', () => {
    it('save to read later', () => {
      const toggleFavorite = jest.fn()
      mockUseFavorites.mockImplementationOnce(() => ({
        favorites: [],
        toggleFavorite,
      }))

      const { getByTestId } = render(
        <NotificationProvider>
          <Story />
        </NotificationProvider>,
      )
      fireEvent.press(getByTestId('button-bookmark'))
      expect(toggleFavorite).toBeCalledWith('id')
    })

    it('remove from read later', () => {
      const toggleFavorite = jest.fn()
      mockUseFavorites.mockImplementationOnce(() => ({
        favorites: ['id'],
        toggleFavorite,
      }))

      const { getByTestId } = render(<Story />)
      fireEvent.press(getByTestId('button-bookmark'))
      expect(toggleFavorite).toBeCalledWith('id')
    })

    it('navigate to read later', () => {
      mockUseFavorites.mockReturnValueOnce({
        favorites: [],
        toggleFavorite: jest.fn(),
      })
      const navigate = jest.fn()
      mockUseNavigation.mockReturnValue({ navigate, addListener: jest.fn() })

      const { getByTestId, getByText } = render(
        <NotificationProvider>
          <Story />
        </NotificationProvider>,
      )
      fireEvent.press(getByTestId('button-bookmark'))
      fireEvent.press(getByText('Ir a leer luego'))
      expect(navigate).toBeCalledWith('Favorite')
    })
  })

  describe('show error message when there is no news', () => {
    it('show error message when service fail', async () => {
      mockUseNavigation.mockImplementationOnce(() => ({ setParams: jest.fn() }))
      mockUseRoute.mockImplementationOnce(() => ({
        params: { pathname: 'pathname' },
      }))
      const spy = jest.spyOn(Alert, 'alert')

      const { queryByTestId } = render(<Story />)
      await act(async () => MockAxios.mockError({ message: '404' }))

      const [title, , buttons] = spy.mock.calls[0]
      buttons?.[0].onPress?.()

      expect(title).toBe('Nota perdida')
      expect(queryByTestId('story-webview')).toBeNull()
    })
  })
})
