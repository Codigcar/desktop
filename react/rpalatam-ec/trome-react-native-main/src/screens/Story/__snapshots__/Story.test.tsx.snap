// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story screen snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {},
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "display": "flex",
          "flex": 1,
        },
      ]
    }
    testID="story-screen"
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderBottomColor": "#FFCB05",
            "borderBottomWidth": 1,
            "display": "flex",
            "elevation": 3,
            "shadowColor": "#050505",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.16,
            "shadowRadius": 2.22,
            "zIndex": 1,
          },
          Object {
            "paddingTop": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "height": 48,
              "justifyContent": "center",
              "paddingHorizontal": 8,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "display": "flex",
                "left": 8,
                "position": "absolute",
                "zIndex": 1,
              },
            ]
          }
        >
          <RNGestureHandlerButton
            collapsable={false}
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            testID="goback-button"
          >
            <View
              accessible={true}
              collapsable={false}
              nativeID="animatedComponent"
              style={Object {}}
            >
              <Icon
                color="#2e3443"
                name="chevron-left"
                size={32}
              />
            </View>
          </RNGestureHandlerButton>
        </View>
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#2e3443",
                "display": "flex",
                "fontFamily": "Avenir",
                "fontSize": 16,
                "fontWeight": "700",
                "lineHeight": 24,
                "textAlign": "center",
              },
              Array [
                Object {
                  "flex": 1,
                  "maxWidth": 654,
                },
                undefined,
              ],
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "display": "flex",
                "position": "absolute",
                "right": 8,
                "zIndex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                },
              ]
            }
            testID="options"
          >
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "marginRight": 8,
                  },
                ]
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="button-bookmark"
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "display": "flex",
                        "height": 32,
                        "justifyContent": "center",
                        "width": 32,
                      },
                    ]
                  }
                >
                  <Icon
                    color="#2e3443"
                    name="bookmark"
                    size={24}
                  />
                </View>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "marginRight": 8,
                  },
                ]
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="button-share"
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "display": "flex",
                        "height": 32,
                        "justifyContent": "center",
                        "width": 32,
                      },
                    ]
                  }
                >
                  <Icon
                    color="#2e3443"
                    name="share-variant"
                    size={24}
                  />
                </View>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                  },
                ]
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="button-font_size"
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "display": "flex",
                        "flexDirection": "row",
                        "height": 32,
                        "justifyContent": "center",
                        "width": 32,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#2e3443",
                          "display": "flex",
                          "fontFamily": "Avenir",
                          "fontSize": 20,
                          "fontWeight": "700",
                          "lineHeight": 28,
                        },
                      ]
                    }
                  >
                    A
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#2e3443",
                          "display": "flex",
                          "fontFamily": "Avenir",
                          "fontSize": 16,
                          "fontWeight": "700",
                          "lineHeight": 36,
                        },
                      ]
                    }
                  >
                    a
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <RNCWebView
        allowsFullscreenVideo={true}
        allowsInlineMediaPlayback={true}
        bounces={false}
        cacheEnabled={true}
        injectedJavaScriptBeforeContentLoaded="
    document.cookie = 'eeecappuser=undefined;'
    window.__REACT_NATIVE_CONTEXT_TOPICS__ = true;
    window.__REACT_NATIVE_SCREENS_AUTH_SIGNWALL__ = false;
    window.__REACT_NATIVE_SCREENS_STORY_V2__ = false;
    window.__REACT_NATIVE_SCREENS_PAYWALL_MODAL__ = false;
    window.sessionStorage.setItem('config', '{\\"fontSize\\":\\"m\\",\\"mode\\":false}');
    true;
  "
        injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
        injectedJavaScriptForMainFrameOnly={true}
        javaScriptEnabled={true}
        mediaPlaybackRequiresUserAction={true}
        messagingEnabled={true}
        onContentProcessDidTerminate={[Function]}
        onHttpError={[Function]}
        onLoadingError={[Function]}
        onLoadingFinish={[Function]}
        onLoadingProgress={[Function]}
        onLoadingStart={[Function]}
        onMessage={[Function]}
        onShouldStartLoadWithRequest={[Function]}
        setSupportMultipleWindows={false}
        source={
          Object {
            "uri": "https://pwa.dev.elcomercio.pe/story/id#?version=50",
          }
        }
        style={
          Array [
            Object {
              "flex": 1,
              "overflow": "hidden",
            },
            Object {
              "backgroundColor": "#ffffff",
            },
            Object {
              "backgroundColor": "#FFFFFF",
              "flex": 1,
            },
          ]
        }
        testID="story-webview"
        textInteractionEnabled={true}
        useSharedProcessPool={true}
      />
    </View>
  </View>
</RNCSafeAreaProvider>
`;
